### PR TITLE
application: Support for more source and policy combinations

### PR DESCRIPTION
**Pull request by Xieql**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

part of #315 

The target of this pull request is to:
Support combination of multiple kinds of `sources` and `syncPolicies`.

To accomplish the above targets, this pull request:
1 Removes the restriction and default value for the combination of `sources` and `syncPolicies`. Add a new example with `GitRepostory` + `HelmRelease`

2 Implements some robust actions. such as handle update conflict.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
can you update the pr title to be more specific?

File: examples/application/quickstart3.yaml
Comment (hzxuzhonghu) [inline examples/application/quickstart3.yaml:1]: if it is for usage doc, can you upload it with docs pr

File: pkg/fleet-manager/application_controller.go [hunk 1]
```diff
@@ -60,9 +60,8 @@ type ApplicationManager struct {
 	Scheme *runtime.Scheme
 }
 
-var fleetToApplicationMap = map[string][]string{}
-var clusterToApplicationMap = map[string][]string{}
-var attachedClusterToApplicationMap = map[string][]string{}
+// relations represent the actual associated resources of the current application.
+var relations = NewRelations()
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/application_controller.go:64]: nit: relations is not a good name, not clear enough

File: examples/application/quickstart3.yaml
Comment (Xieql) [reply examples/application/quickstart3.yaml:1]: yes, i will add it

**Comment by hzxuzhonghu** [comment]
After application api update, you may need to change accordingly

**Comment by Xieql** [comment]
> After application api update, you may need to change accordingly

ok

**Comment by Xieql** [comment]
/ hold

File: examples/application/quickstart3.yaml
Comment (hzxuzhonghu) [inline examples/application/quickstart3.yaml:1]: I mean this seems not neccessary to be within this pr

File: pkg/fleet-manager/application_helper.go [hunk 2]
```diff
@@ -457,8 +491,8 @@ func generateKustomizationName(app *applicationapi.Application, clusterKind, clu
 }
 
 // generateHelmReleaseName constructs a unique name for HelmRelease based on the provided application,
-func generateHelmReleaseName(app *applicationapi.Application, clusterKind, clusterName, policyName string) string {
-	name := app.Name + "-" + policyName + "-" + clusterKind + "-" + clusterName
+func generateHelmReleaseName(policyName, clusterKind, clusterName string) string {
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/application_helper.go:494]: is this duplicate with generateKustomizationName

File: pkg/fleet-manager/application_helper.go [hunk 3]
```diff
@@ -442,13 +469,20 @@ func findSourceKind(app *applicationapi.Application) string {
 	return ""
 }
 
-// generateKustomizationName constructs a unique name for Kustomization based on the provided application,
-// synchronization policy, cluster kind and cluster name. The resulting name is formatted to be lower-case,
-// and is truncated to a maximum of 63 characters if needed.
+// findPolicyKind get the type of the application's syncPolicy.
+func findPolicyKind(syncPolicy *applicationapi.ApplicationSyncPolicy) string {
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/application_helper.go:473]: nit: `getSyncPolicyKind`

File: pkg/fleet-manager/application_helper.go [hunk 4]
```diff
@@ -37,101 +38,121 @@ import (
...
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to fetch kustomizationList for application: %w", err)
+	}
+	return helmReleaseList, nil
+}
 
-	var clusterList clusterv1alpha1.ClusterList
-	var attachedClusterList clusterv1alpha1.AttachedClusterList
+// handleSyncByPolicyKind handles syncing for a given policy kind (either kustomization or Helm release) based on the provided sync policy.
+func (a *ApplicationManager) handleSyncPolicyByKind(ctx context.Context, app *applicationapi.Application, policyKind string, syncPolicy *applicationapi.ApplicationSyncPolicy, policyName string, fleetCluster ClusterInterface, kubeConfig *fluxmeta.KubeConfigReference) (ctrl.Result, error) {
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/application_helper.go:131]: This is too long, better make each arg a line

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**